[flex_attention]: adding a heuristic for USE_TMA

dhruvak3 opened the pull request:
Summary:
Adding a USE_TMA heuristic which captures almost all of the perf.

The best perf improvements are seen in document_mask+large shapes which are very important for training.

Mean: +15.8% improvement overall

Only a single shape (below) sees a perf degradation of >5%.

Document Mask: 37.8% average improvement
* Strong performance on variable-length sequences
* Once case of regression on a short sequences (-5.72%)

Sliding Window: 58.7% average improvement
* 30-97% improvements

Head Bias: 2.5% average improvement
* Consistent 2-3% gains across all sequence lengths

Others:
* Softcap: 0.15% average improvement (minimal impact)
* Prefix_LM: -0.4% average change (slight degradation)
* Full Attention (noop): 1.6% average improvement
* Causal: 0.6% average improvement

Note: sliding window uplift is because of the test side change.

Test Plan:
Ran 
```
buck2 run @//mode/opt -c fbcode.nvcc_arch=b200a -c fbcode.enable_gpu_sections=true -c fbcode.platform010_cuda_version=12.8 //pytorch/tritonbench:run -- --op flex_attention --metrics tflops,latency
```

Across the different variations.

Here is the uplift plot:
 {F1982793932}

Raw Data:
```
Shape	Mask Type	Post Changes	Pre-Changes	Percent Change
(8, 16, 128, 16, 128, 128)	noop	12.2404	11.672	4.869773818
(8, 16, 128, 16, 128, 128)	causal	19.6761	18.9992	3.562781591
(8, 16, 128, 16, 128, 128)	rel	12.64	12.1127	4.353282092
(8, 16, 128, 16, 128, 128)	head_bias	12.3075	12.0091	2.484782373
(8, 16, 128, 16, 128, 128)	alibi	19.0314	18.2775	4.124743537
(8, 16, 128, 16, 128, 128)	sliding_window	19.1613	18.5902	3.072048714
(1, 16, 1024, 16, 1024, 128)	document_mask	28.8403	30.5914	-5.72415777
(8, 16, 128, 16, 128, 128)	prefix_lm	18.3073	18.8398	-2.826463126
(8, 16, 128, 16, 128, 128)	softcap	18.2974	18.3773	-0.4347755111
(8, 16, 256, 16, 256, 128)	noop	48.6082	47.8148	1.659318872
(8, 16, 256, 16, 256, 128)	causal	57.6149	56.2675	2.394632781
(8, 16, 256, 16, 256, 128)	rel	47.4945	47.2779	0.4581421764
(8, 16, 256, 16, 256, 128)	head_bias	48.3813	47.014	2.908282639
(8, 16, 256, 16, 256, 128)	alibi	56.9332	54.2731	4.901323123
(8, 16, 256, 16, 256, 128)	sliding_window	56.4246	52.6885	7.090921169
(1, 16, 2048, 16, 2048, 128)	document_mask	76.3089	74.3736	2.602133015
(8, 16, 256, 16, 256, 128)	prefix_lm	58.2794	55.5256	4.95951417
(8, 16, 256, 16, 256, 128)	softcap	55.2826	53.8968	2.571210165
(8, 16, 512, 16, 512, 128)	noop	183.26	187.142	-2.074360646
(8, 16, 512, 16, 512, 128)	causal	174.899	172.658	1.297941596
(8, 16, 512, 16, 512, 128)	rel	189.178	185.085	2.211416376
(8, 16, 512, 16, 512, 128)	head_bias	187.142	181.349	3.194393131
(8, 16, 512, 16, 512, 128)	alibi	164.641	164.48	0.09788424125
(8, 16, 512, 16, 512, 128)	sliding_window	169.957	122.94	38.24385879
(1, 16, 4096, 16, 4096, 128)	document_mask	146.112	107.093	36.43468761
(8, 16, 512, 16, 512, 128)	prefix_lm	175.354	175.354	0
(8, 16, 512, 16, 512, 128)	softcap	162.262	162.183	0.04871040738
(8, 16, 1024, 16, 1024, 128)	noop	451.681	446.073	1.257193329
(8, 16, 1024, 16, 1024, 128)	causal	252.694	252.641	0.02097838435
(8, 16, 1024, 16, 1024, 128)	rel	446.166	443.139	0.6830813808
(8, 16, 1024, 16, 1024, 128)	head_bias	454.921	446.258	1.941253714
(8, 16, 1024, 16, 1024, 128)	alibi	236.666	236.712	-0.01943289736
(8, 16, 1024, 16, 1024, 128)	sliding_window	249.215	167.812	48.50844993
(1, 16, 8192, 16, 8192, 128)	document_mask	210.169	149.254	40.81297654
(8, 16, 1024, 16, 1024, 128)	prefix_lm	254.604	254.284	0.125843545
(8, 16, 1024, 16, 1024, 128)	softcap	229.892	229.848	0.01914308587
(8, 16, 2048, 16, 2048, 128)	noop	608.016	597.526	1.755572142
(8, 16, 2048, 16, 2048, 128)	causal	321.017	321.693	-0.2101382374
(8, 16, 2048, 16, 2048, 128)	rel	587.111	586.792	0.054363386
(8, 16, 2048, 16, 2048, 128)	head_bias	598.687	587.111	1.971688488
(8, 16, 2048, 16, 2048, 128)	alibi	295.773	295.773	0
(8, 16, 2048, 16, 2048, 128)	sliding_window	318.097	183.105	73.72381967
(1, 16, 16384, 16, 16384, 128)	document_mask	285.551	195.233	46.26164634
(8, 16, 2048, 16, 2048, 128)	prefix_lm	322.304	321.806	0.1547516205
(8, 16, 2048, 16, 2048, 128)	softcap	286.219	285.88	0.1185812229
(8, 16, 4096, 16, 4096, 128)	noop	623.445	625.253	-0.2891629468
(8, 16, 4096, 16, 4096, 128)	causal	369.057	369.057	0
(8, 16, 4096, 16, 4096, 128)	rel	626.718	630.683	-0.6286835066
(8, 16, 4096, 16, 4096, 128)	head_bias	619.192	617.717	0.2387824845
(8, 16, 4096, 16, 4096, 128)	alibi	336.842	336.849	-0.002078082464
(8, 16, 4096, 16, 4096, 128)	sliding_window	367.357	192.262	91.07103848
(1, 16, 32768, 16, 32768, 128)	document_mask	352.813	236.054	49.46283478
(8, 16, 4096, 16, 4096, 128)	prefix_lm	369.309	369.309	0
(8, 16, 4096, 16, 4096, 128)	softcap	324.088	324.112	-0.007404847707
(8, 16, 8192, 16, 8192, 128)	noop	668.267	644.41	3.702146149
(8, 16, 8192, 16, 8192, 128)	causal	393.945	397.241	-0.8297230145
(8, 16, 8192, 16, 8192, 128)	rel	663.352	649.078	2.199119366
(8, 16, 8192, 16, 8192, 128)	head_bias	653.654	654.78	-0.1719661566
(8, 16, 8192, 16, 8192, 128)	alibi	363.893	360.898	0.8298743689
(8, 16, 8192, 16, 8192, 128)	sliding_window	385.163	197.756	94.76678331
(1, 16, 65536, 16, 65536, 128)	document_mask	405.668	264.619	53.3026729
(8, 16, 8192, 16, 8192, 128)	prefix_lm	393.872	397.458	-0.9022336951
(8, 16, 8192, 16, 8192, 128)	softcap	345.572	346.346	-0.223475946
(8, 16, 16384, 16, 16384, 128)	noop	681.404	642.487	6.057243182
(8, 16, 16384, 16, 16384, 128)	causal	416.768	413.811	0.7145774279
(8, 16, 16384, 16, 16384, 128)	rel	676.936	660.778	2.445299329
(8, 16, 16384, 16, 16384, 128)	head_bias	676.907	657.428	2.962910007
(8, 16, 16384, 16, 16384, 128)	alibi	382.453	373.863	2.29763309
(8, 16, 16384, 16, 16384, 128)	sliding_window	395.434	200.448	97.27510377
(1, 16, 131072, 16, 131072, 128)	document_mask	439.601	282.557	55.57958217
(8, 16, 16384, 16, 16384, 128)	prefix_lm	416.927	412.476	1.079093087
(8, 16, 16384, 16, 16384, 128)	softcap	363.194	359.052	1.153593351
```

Differential Revision: D84373821




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

Skylion007 commented on torch/_inductor/kernel/flex/flex_attention.py:
```diff
@@ -342,8 +342,16 @@ def flex_attention(
-        # USE TMA = false by default
-        cur_kernel_options.setdefault("USE_TMA", False)
+        # USE TMA = false by default, except for document_mask when B=1 or when M/N >= 8192
+        batch_size = V.graph.sizevars.evaluate_expr(Bq)
```
Why not just have AUTOTUNE toggle the use TMA flag as part of torch compile?

dhruvak3 committed:
[flex_attention]: adding a heuristic for USE_TMA (#165800)

Summary:

Adding a USE_TMA heuristic which captures almost all of the perf.

The best perf improvements are seen in document_mask+large shapes which are very important for training.

Mean: +15.8% improvement overall

Only a single shape (below) sees a perf degradation of >5%.

Document Mask: 37.8% average improvement
* Strong performance on variable-length sequences
* Once case of regression on a short sequences (-5.72%)

Sliding Window: 58.7% average improvement
* 30-97% improvements

Head Bias: 2.5% average improvement
* Consistent 2-3% gains across all sequence lengths

Others:
* Softcap: 0.15% average improvement (minimal impact)
* Prefix_LM: -0.4% average change (slight degradation)
* Full Attention (noop): 1.6% average improvement
* Causal: 0.6% average improvement

Note: sliding window uplift is because of the test side change.

Test Plan:
Ran 
```
buck2 run @//mode/opt -c fbcode.nvcc_arch=b200a -c fbcode.enable_gpu_sections=true -c fbcode.platform010_cuda_version=12.8 //pytorch/tritonbench:run -- --op flex_attention --metrics tflops,latency
```

Across the different variations.

Here is the uplift plot:
 {F1982793932}

Raw Data:
```
Shape	Mask Type	Post Changes	Pre-Changes	Percent Change
(8, 16, 128, 16, 128, 128)	noop	12.2404	11.672	4.869773818
(8, 16, 128, 16, 128, 128)	causal	19.6761	18.9992	3.562781591
(8, 16, 128, 16, 128, 128)	rel	12.64	12.1127	4.353282092
(8, 16, 128, 16, 128, 128)	head_bias	12.3075	12.0091	2.484782373
(8, 16, 128, 16, 128, 128)	alibi	19.0314	18.2775	4.124743537
(8, 16, 128, 16, 128, 128)	sliding_window	19.1613	18.5902	3.072048714
(1, 16, 1024, 16, 1024, 128)	document_mask	28.8403	30.5914	-5.72415777
(8, 16, 128, 16, 128, 128)	prefix_lm	18.3073	18.8398	-2.826463126
(8, 16, 128, 16, 128, 128)	softcap	18.2974	18.3773	-0.4347755111
(8, 16, 256, 16, 256, 128)	noop	48.6082	47.8148	1.659318872
(8, 16, 256, 16, 256, 128)	causal	57.6149	56.2675	2.394632781
(8, 16, 256, 16, 256, 128)	rel	47.4945	47.2779	0.4581421764
(8, 16, 256, 16, 256, 128)	head_bias	48.3813	47.014	2.908282639
(8, 16, 256, 16, 256, 128)	alibi	56.9332	54.2731	4.901323123
(8, 16, 256, 16, 256, 128)	sliding_window	56.4246	52.6885	7.090921169
(1, 16, 2048, 16, 2048, 128)	document_mask	76.3089	74.3736	2.602133015
(8, 16, 256, 16, 256, 128)	prefix_lm	58.2794	55.5256	4.95951417
(8, 16, 256, 16, 256, 128)	softcap	55.2826	53.8968	2.571210165
(8, 16, 512, 16, 512, 128)	noop	183.26	187.142	-2.074360646
(8, 16, 512, 16, 512, 128)	causal	174.899	172.658	1.297941596
(8, 16, 512, 16, 512, 128)	rel	189.178	185.085	2.211416376
(8, 16, 512, 16, 512, 128)	head_bias	187.142	181.349	3.194393131
(8, 16, 512, 16, 512, 128)	alibi	164.641	164.48	0.09788424125
(8, 16, 512, 16, 512, 128)	sliding_window	169.957	122.94	38.24385879
(1, 16, 4096, 16, 4096, 128)	document_mask	146.112	107.093	36.43468761
(8, 16, 512, 16, 512, 128)	prefix_lm	175.354	175.354	0
(8, 16, 512, 16, 512, 128)	softcap	162.262	162.183	0.04871040738
(8, 16, 1024, 16, 1024, 128)	noop	451.681	446.073	1.257193329
(8, 16, 1024, 16, 1024, 128)	causal	252.694	252.641	0.02097838435
(8, 16, 1024, 16, 1024, 128)	rel	446.166	443.139	0.6830813808
(8, 16, 1024, 16, 1024, 128)	head_bias	454.921	446.258	1.941253714
(8, 16, 1024, 16, 1024, 128)	alibi	236.666	236.712	-0.01943289736
(8, 16, 1024, 16, 1024, 128)	sliding_window	249.215	167.812	48.50844993
(1, 16, 8192, 16, 8192, 128)	document_mask	210.169	149.254	40.81297654
(8, 16, 1024, 16, 1024, 128)	prefix_lm	254.604	254.284	0.125843545
(8, 16, 1024, 16, 1024, 128)	softcap	229.892	229.848	0.01914308587
(8, 16, 2048, 16, 2048, 128)	noop	608.016	597.526	1.755572142
(8, 16, 2048, 16, 2048, 128)	causal	321.017	321.693	-0.2101382374
(8, 16, 2048, 16, 2048, 128)	rel	587.111	586.792	0.054363386
(8, 16, 2048, 16, 2048, 128)	head_bias	598.687	587.111	1.971688488
(8, 16, 2048, 16, 2048, 128)	alibi	295.773	295.773	0
(8, 16, 2048, 16, 2048, 128)	sliding_window	318.097	183.105	73.72381967
(1, 16, 16384, 16, 16384, 128)	document_mask	285.551	195.233	46.26164634
(8, 16, 2048, 16, 2048, 128)	prefix_lm	322.304	321.806	0.1547516205
(8, 16, 2048, 16, 2048, 128)	softcap	286.219	285.88	0.1185812229
(8, 16, 4096, 16, 4096, 128)	noop	623.445	625.253	-0.2891629468
(8, 16, 4096, 16, 4096, 128)	causal	369.057	369.057	0
(8, 16, 4096, 16, 4096, 128)	rel	626.718	630.683	-0.6286835066
(8, 16, 4096, 16, 4096, 128)	head_bias	619.192	617.717	0.2387824845
(8, 16, 4096, 16, 4096, 128)	alibi	336.842	336.849	-0.002078082464
(8, 16, 4096, 16, 4096, 128)	sliding_window	367.357	192.262	91.07103848
(1, 16, 32768, 16, 32768, 128)	document_mask	352.813	236.054	49.46283478
(8, 16, 4096, 16, 4096, 128)	prefix_lm	369.309	369.309	0
(8, 16, 4096, 16, 4096, 128)	softcap	324.088	324.112	-0.007404847707
(8, 16, 8192, 16, 8192, 128)	noop	668.267	644.41	3.702146149
(8, 16, 8192, 16, 8192, 128)	causal	393.945	397.241	-0.8297230145
(8, 16, 8192, 16, 8192, 128)	rel	663.352	649.078	2.199119366
(8, 16, 8192, 16, 8192, 128)	head_bias	653.654	654.78	-0.1719661566
(8, 16, 8192, 16, 8192, 128)	alibi	363.893	360.898	0.8298743689
(8, 16, 8192, 16, 8192, 128)	sliding_window	385.163	197.756	94.76678331
(1, 16, 65536, 16, 65536, 128)	document_mask	405.668	264.619	53.3026729
(8, 16, 8192, 16, 8192, 128)	prefix_lm	393.872	397.458	-0.9022336951
(8, 16, 8192, 16, 8192, 128)	softcap	345.572	346.346	-0.223475946
(8, 16, 16384, 16, 16384, 128)	noop	681.404	642.487	6.057243182
(8, 16, 16384, 16, 16384, 128)	causal	416.768	413.811	0.7145774279
(8, 16, 16384, 16, 16384, 128)	rel	676.936	660.778	2.445299329
(8, 16, 16384, 16, 16384, 128)	head_bias	676.907	657.428	2.962910007
(8, 16, 16384, 16, 16384, 128)	alibi	382.453	373.863	2.29763309
(8, 16, 16384, 16, 16384, 128)	sliding_window	395.434	200.448	97.27510377
(1, 16, 131072, 16, 131072, 128)	document_mask	439.601	282.557	55.57958217
(8, 16, 16384, 16, 16384, 128)	prefix_lm	416.927	412.476	1.079093087
(8, 16, 16384, 16, 16384, 128)	softcap	363.194	359.052	1.153593351
```

Differential Revision: D84373821
